contracts/Lockbox.cdc

```diff
@@ -0,0 +1,468 @@
+
+            let vault <- vaultRef.withdraw(amount: amount)
+
+            self.decreaseUnlockLimit(delta: amount)
+
+            return <- vault
+        }
+
+        access(self) fun decreaseUnlockLimit(delta: UFix64) {  
+            self.unlockLimit = self.unlockLimit - delta
+        }
+
+        // Lockbox.TokenAdmin actions
+
+        /// Called by the admin every time a vesting release happens
+        pub fun increaseUnlockLimit(delta: UFix64) {  
+            self.unlockLimit = self.unlockLimit + delta
+        }
+
+        // Lockbox.TokenHolder actions
+
+        /// Registers a new node operator with the Flow Staking contract
+        /// and commits an initial amount of locked tokens to stake
+        pub fun registerNode(nodeInfo: StakingProxy.NodeInfo, amount: UFix64) {
+            let vaultRef = self.vault.borrow()!
+
+            let tokens <- vaultRef.withdraw(amount: amount)
+
+            self.nodeStaker <-! FlowIDTableStaking.addNodeRecord(id: nodeInfo.id, role: nodeInfo.role, networkingAddress: nodeInfo.networkingAddress, networkingKey: nodeInfo.networkingKey, stakingKey: nodeInfo.stakingKey, tokensCommitted: <-tokens)
+        }
+
+        /// Registers a new Delegator with the Flow Staking contract
+        /// the caller has to specify the ID of the node operator
+        /// they are delegating to
+        pub fun registerDelegator(nodeID: String) {
+            self.nodeDelegator <-! FlowIDTableStaking.registerNewDelegator(nodeID: nodeID)
+        }
+    }
+
+    /// allows anyone to read the unlock limit of an account with TokenHolder
+    pub resource interface UnlockLimit {
+        pub fun getUnlockLimit(): UFix64
+    }
+
+    // Stored in Holder unlocked account
+    pub resource TokenHolder: UnlockLimit {
+
+        /// Capability that is used to access the LockedTokenManager
+        /// in the shared account
+        access(self) var tokenManager: Capability<&LockedTokenManager>
+
+        /// Used to perform staking actions if the user has signed up
+        /// as a node operator
+        access(self) var nodeStakerProxy: LockedNodeStakerProxy?
+
+        /// Used to perform delegating actions if the user has signed up
+        /// as a delegator
+        access(self) var nodeDelegatorProxy: LockedNodeDelegatorProxy?
+
+        init(tokenManager: Capability<&LockedTokenManager>) {
+            pre {
+                tokenManager.borrow() != nil: "Must pass a LockedTokenManager capability"
+            }
+            self.tokenManager = tokenManager
+            self.nodeStakerProxy = nil
+            self.nodeDelegatorProxy = nil
+        }
+
+        pub fun getUnlockLimit(): UFix64 {
+            let tokenManagerRef = self.tokenManager.borrow()!
+
+            return tokenManagerRef.unlockLimit
+        }
+
+        /// Deposits tokens in the locked vault, which marks them as 
+        /// unlocked and available to withdraw
+        pub fun deposit(from: @FungibleToken.Vault) {
```

Comment:
Should `TokenHolder` also implement `FungibleToken.Receiver` and `FungibleToken.Provider`? It might be nice to allow the user to route standard token transfers to the lockbox. The methods are already defined -- we'd just need to add the conformances to this resource defintion.